### PR TITLE
CP-36178: Add basic precheck function for updates

### PR DESCRIPTION
**Pull request by robhoes**
Also includes some minor improvements.

### CODE REVIEW

File: ocaml/xapi/rbac_audit.ml [hunk 1]
```diff
@@ -443,19 +443,24 @@ let has_to_audit action =
           [
             (* list of _actions_ filtered out from the audit log *)
             "session.local_logout"
+          ; "session_local_logout"
```
Comment (lippirk) [inline ocaml/xapi/rbac_audit.ml:446]: why are we duplicating these?

File: ocaml/xapi/xapi_pool.ml [hunk 2]
```diff
@@ -3325,6 +3321,28 @@ let sync_updates ~__context ~self ~force =
       List.iter (fun x -> create_pool_repository ~__context ~self:x) enabled ;
       set_available_updates ~__context
 
+let check_update_readiness ~__context ~self ~requires_reboot =
```
Comment (lippirk) [inline ocaml/xapi/xapi_pool.ml:3324]: by this logic, we consider a pool 'ready to upgrade' if there are no miscreant VMs, and if all the hosts are up

what about if there's a cluster running, or HA is enabled, etc. ?

File: ocaml/xapi/xapi_pool.ml [hunk 3]
```diff
@@ -3325,6 +3321,28 @@ let sync_updates ~__context ~self ~force =
...
+  List.map
```
Comment (lippirk) [inline ocaml/xapi/xapi_pool.ml:3325]:
```suggestion
  List.concat_map
```

sadly there is no `filter_concat_map`

File: ocaml/xapi/message_forwarding.ml [hunk 4]
```diff
@@ -1021,6 +1021,12 @@ functor
           (pool_uuid ~__context self)
           (string_of_bool force) ;
         Local.Pool.sync_updates ~__context ~self ~force
+
+      let check_update_readiness ~__context ~self ~requires_reboot =
+        info "Pool.check_update_readiness: pool = '%s'; requires_reboot = %s"
```
Comment (lippirk) [inline ocaml/xapi/message_forwarding.ml:1026]: you can use %b here

File: ocaml/idl/datamodel_pool.ml [hunk 5]
```diff
@@ -763,6 +748,19 @@ open Datamodel_types
...
 
+  let check_update_readiness = call
+      ~name:"check_update_readiness"
+      ~lifecycle:[Published, rel_next, ""]
+      ~doc:"Check if the pool is ready to be updated. If not, report the reasons."
+      ~params:[
+        Ref _pool, "self", "The pool";
+        Bool, "requires_reboot", "Assume that the update will require host reboots"
+      ]
+      ~result:(Set(Set (String)), "A set of error codes with arguments, if the pool is
```
Comment (lippirk) [inline ocaml/idl/datamodel_pool.ml:759]: as far as I can tell, it should be `Set (Pair String))`? not sure if the datamodel supports that though

Comment (robhoes) [reply ocaml/idl/datamodel_pool.ml:759]: No, it is not a pair, but list: the first element is always the error code and then there are zero or more arguments (as in datamodel_errors.ml). A better type would be `string * string list`, but the datamodel doesn't support tuples.

File: ocaml/xapi/xapi_pool.ml (hunk 2 above)
Comment (robhoes) [reply ocaml/xapi/xapi_pool.ml:3324]: Yeah, it just so happens that our update service already checks for HA and WLB, and disables them proactively. I would still add checks for them for extra safety though.

**Review by minglumlu — APPROVED** [review]

File: ocaml/idl/datamodel_pool.ml (hunk 5 above)
[the remark this answers is not in the document]
Comment (robhoes) [reply ocaml/idl/datamodel_pool.ml:759]: Unfortunately, the datamodel types are not flexible enough to make this much better; see my reply to Ben's comment above.

File: ocaml/xapi/xapi_pool.ml (hunk 2 above)
Comment (robhoes) [reply ocaml/xapi/xapi_pool.ml:3370]: I've added the HA check now. We are not currently convinced that anything else is needed.

File: ocaml/xapi/xapi_pool.ml [hunk 6]
```diff
@@ -3325,6 +3321,28 @@ let sync_updates ~__context ~self ~force =
...
+    (fun host ->
```
[the remark this answers is not in the document]
Comment (robhoes) [reply ocaml/xapi/xapi_pool.ml]: I've made this change now.

**Pull request merged**